File: leetcode/78.py
from typing import List

class Solution:
  def subsets(self, nums:List[int]) -> List[List[int]]:
    def backtrack(k:int, n:int):
      if k == n:
        pos.append(ss.copy())
        return
      backtrack(k+1, n)
      ss.append(nums[k])
      backtrack(k+1, n)
      ss.pop()
    ss = []
    pos = []
    backtrack(0, len(nums))
    return pos

File: leetcode/78.cpp
#include <iostream>
#include <bits/stdc++.h>

class Solution {
  public:
    vector<vector<int>> subsets(vector<int>& nums) {
      vector<int> ss{};
      vector<vector<int>> pos{};
      search(nums, ss, pos, 0);
      return pos;
    }
  private:
    void search(vector<int>& nums, vector<int>& ss, vector<vector<int>>& pos, int k) {
      if (k == nums.size()) {
        pos.push_back(ss);
        return;
      }
      ss.push_back(nums[k]);
      search(nums, ss, pos, k+1);
      ss.pop_back();
      search(nums, ss, pos, k+1);
    }
};

